### PR TITLE
IdentityProviderConfig persistence layer [DPP-1331]

### PR DESCRIPTION
**Pull request by skisel-da**

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/localstore/PersistentIdentityProviderConfigStore.scala [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+  private def idpConfigDoesNotExist(
+      id: IdentityProviderId.Id
+  )(implicit connection: Connection): Result[Unit] = Either.cond(
+    !backend.idpConfigByIdExists(id)(connection),
+    (),
+    IdentityProviderConfigExists(id),
+  )
+
+  private def idpConfigIssuerDoesNotExist(
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/localstore/PersistentIdentityProviderConfigStore.scala:132]:
```suggestion
  private def idpConfigByIssuerDoesNotExist(
```
?

File: ledger/participant-integration-api/src/main/scala/platform/localstore/PersistentIdentityProviderConfigStore.scala [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    extends IdentityProviderConfigStore {
+
+  private val backend = dbSupport.storageBackendFactory.createIdentityProviderConfigStorageBackend
+  private val dbDispatcher = dbSupport.dbDispatcher
+  private val logger = ContextualizedLogger.get(getClass)
+
+  override def createIdentityProviderConfig(identityProviderConfig: domain.IdentityProviderConfig)(
+      implicit loggingContext: LoggingContext
+  ): Future[Result[domain.IdentityProviderConfig]] = {
+    inTransaction(_.createIDPConfig) { implicit connection =>
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/localstore/PersistentIdentityProviderConfigStore.scala:35]: Consider using helper methods `wihtIdpConfig` and `withoutIdpConfig` similar to `withUser` and `withoutUser` in user management

File: ledger/participant-integration-api/src/main/scala/platform/localstore/PersistentIdentityProviderConfigStore.scala [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      dbSupport: DbSupport,
+      metrics: Metrics,
+      expiryAfterWriteInSeconds: Int,
+      maximumCacheSize: Int,
+  )(implicit
+      executionContext: ExecutionContext,
+      loggingContext: LoggingContext,
+  ) = new CachedIdentityProviderConfigStore(
+    new PersistentIdentityProviderConfigStore(dbSupport, metrics),
+    expiryAfterWriteInSeconds,
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/localstore/PersistentIdentityProviderConfigStore.scala:168]: Use named arguments as we are passing down two int values

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderCheckStorageBackend.scala [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.daml.platform.store.backend.localstore
+
+import com.daml.ledger.api.domain.IdentityProviderId
+
+import java.sql.Connection
+
+trait IdentityProviderCheckStorageBackend {
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderCheckStorageBackend.scala:9]: Why did you create a separate backend with just this one method?

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderCheckStorageBackendImpl.scala [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+import anorm.RowParser
+import anorm.SqlParser.int
+import com.daml.ledger.api.domain.IdentityProviderId
+import com.daml.platform.store.backend.common.SimpleSqlAsVectorOf._
+
+import java.sql.Connection
+
+object IdentityProviderCheckStorageBackendImpl extends IdentityProviderCheckStorageBackend {
+  private val IntParser0: RowParser[Int] =
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderCheckStorageBackendImpl.scala:13]:
```suggestion
  private val IntParser: RowParser[Int] =
```

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderCheckStorageBackendImpl.scala [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    int("dummy") map { i => i }
+
+  def idpConfigByIdExists(id: IdentityProviderId.Id)(connection: Connection): Boolean = {
+    import com.daml.platform.store.backend.common.ComposableQuery.SqlStringInterpolation
+    val res: Seq[_] =
+      SQL"""
+         SELECT 1 AS dummy
+         FROM participant_identity_provider_config idpcfg
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderCheckStorageBackendImpl.scala:21]:
```suggestion
         FROM participant_identity_provider_config t
```
?
`idpcfg` is hard to parse

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderStorageBackend.scala [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+  def updateJwksURL(id: IdentityProviderId.Id, jwksURL: JwksUrl)(
+      connection: Connection
+  ): Boolean
+
+  def updateIsDeactivated(id: IdentityProviderId.Id, isDeactivated: Boolean)(
+      connection: Connection
+  ): Boolean
+
+  def idpConfigByIssuerExists(issuer: String)(connection: Connection): Boolean
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderStorageBackend.scala:37]:
```suggestion
  def identityProviderConfigByIssuerExists(issuer: String)(connection: Connection): Boolean
```
for consistency

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderStorageBackend.scala [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+  def listIdentityProviderConfigs()(
+      connection: Connection
+  ): Vector[IdentityProviderConfig]
+
+  def updateIssuer(id: IdentityProviderId.Id, newIssuer: String)(
+      connection: Connection
+  ): Boolean
+
+  def updateJwksURL(id: IdentityProviderId.Id, jwksURL: JwksUrl)(
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderStorageBackend.scala:29]:
```suggestion
  def updateJwksUrl(id: IdentityProviderId.Id, jwksUrl: JwksUrl)(
```

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderStorageBackend.scala [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+  def updateIsDeactivated(id: IdentityProviderId.Id, isDeactivated: Boolean)(
+      connection: Connection
+  ): Boolean
+
+  def idpConfigByIssuerExists(issuer: String)(connection: Connection): Boolean
+
+}
+
+object IdentityProviderStorageBackend {}
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderStorageBackend.scala:41]: Is this needed?

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderStorageBackendImpl.scala [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+import com.daml.scalautil.Statement.discard
+
+import java.sql.Connection
+
+object IdentityProviderStorageBackendImpl extends IdentityProviderStorageBackend {
+
+  private val IntParser0: RowParser[Int] =
+    int("dummy") map { i => i }
+
+  private val IDPConfigRecordParser: RowParser[(String, Boolean, String, String)] = {
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderStorageBackendImpl.scala:19]:
```suggestion
  private val IdpConfigRecordParser: RowParser[(String, Boolean, String, String)] = {
```

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderStorageBackendImpl.scala [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+        )
+      }
+  }
+
+  override def idpConfigByIssuerExists(issuer: String)(connection: Connection): Boolean = {
+    import com.daml.platform.store.backend.common.ComposableQuery.SqlStringInterpolation
+    val res: Seq[_] =
+      SQL"""
+         SELECT 1 AS dummy
+         FROM participant_identity_provider_config idpcfg
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderStorageBackendImpl.scala:94]:
```suggestion
         FROM participant_identity_provider_config t
```

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderStorageBackendImpl.scala [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    rowsUpdated == 1
+  }
+
+  override def updateJwksURL(id: IdentityProviderId.Id, jwksURL: JwksUrl)(
+      connection: Connection
+  ): Boolean = {
+    val rowsUpdated =
+      SQL"""
+         UPDATE participant_identity_provider_config
+         SET jwks_url  = ${jwksURL.value}
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderStorageBackendImpl.scala:122]:
```suggestion
         SET jwks_url = ${jwksURL.value}
```

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsIDPConfig.scala [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    val cfg1 = config()
+    val cfg2 = config()
+    val cfg3 = config()
+    executeSql(tested.createIdentityProviderConfig(cfg1))
+    executeSql(tested.createIdentityProviderConfig(cfg2))
+    executeSql(tested.createIdentityProviderConfig(cfg3))
+
+    executeSql(
+      tested.listIdentityProviderConfigs()
+    ) should contain theSameElementsAs Vector(cfg1, cfg2, cfg3)
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsIDPConfig.scala:148]: afaict you're not checking the order of elements why the test case's description claims order by id

File: ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/CachedIdentityProviderConfigStore.scala [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        .maximumSize(maximumCacheSize.toLong)
+        .buildAsync(
+          new FutureAsyncCacheLoader[IdentityProviderId.Id, Result[IdentityProviderConfig]](
+            delegate.getIdentityProviderConfig
+          )
+        ),
+      metrics.daml.identityProviderConfigStore.cache,
+    )
+
+  private val idpListCache: CaffeineCache.AsyncLoadingCaffeineCache[
```
Comment (pbatko-da) [inline ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/CachedIdentityProviderConfigStore.scala:44]: Why do you need two caches?
Is it because the list of all idps will be actively used by authorization on every call?
If it's just for the LAPI rpc then we might be better of dropping it

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres/V126__identity_provider_config.sql [hunk 15]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE participant_identity_provider_config
+(
+    identity_provider_id VARCHAR(255) PRIMARY KEY NOT NULL,
+    issuer               VARCHAR             NOT NULL UNIQUE,
+    jwks_url             VARCHAR             NOT NULL,
+    is_deactivated       BOOLEAN             NOT NULL
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/resources/db/migration/postgres/V126__identity_provider_config.sql:6]: Consider adding ` created_at          NUMBER          NOT NULL`.
It would be unused the same as user's created_at but it might useful when debugging an issue in a live system

File: ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/InMemoryIdentityProviderConfigStore.scala [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+  override def updateIdentityProviderConfig(update: IdentityProviderConfigUpdate)(implicit
+      loggingContext: LoggingContext
+  ): Future[Result[IdentityProviderConfig]] = withState {
+    val id = update.identityProviderId
+    for {
+      currentState <- checkIdExists(id)
+      _ <- update.issuerUpdate.map(checkIssuerDoNotExists).getOrElse(Right(()))
+    } yield {
+      currentState
```
Comment (pbatko-da) [inline ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/InMemoryIdentityProviderConfigStore.scala:67]: Don't you need to insert the updated value somewhere?

File: ledger/participant-local-store/src/test/lib/com/daml/platform/localstore/IdentityProviderConfigStoreTests.scala [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+  def randomId() = {
+    val id = UUID.randomUUID().toString
+    IdentityProviderId.Id(Ref.LedgerString.assertFromString(id))
+  }
+
+  "identity provider config store" - {
+    "allows to create and load unchanged an identity provider config" in {
+      testIt { tested =>
+        val cfg1 = config()
+        val cfg2 = config()
```
Comment (pbatko-da) [inline ledger/participant-local-store/src/test/lib/com/daml/platform/localstore/IdentityProviderConfigStoreTests.scala:40]: What's the difference in `cfg1` and `cfg2`? Can we this test just one of them?

File: ledger/participant-local-store/src/test/lib/com/daml/platform/localstore/IdentityProviderConfigStoreTests.scala [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+      testIt { tested =>
+        val cfg1 = config().copy(issuer = "issuer1")
+        val cfg2 = config().copy(issuer = "issuer2")
+        for {
+          _ <- tested.createIdentityProviderConfig(cfg1)
+          _ <- tested.createIdentityProviderConfig(cfg2)
+          res <- tested.updateIdentityProviderConfig(
+            IdentityProviderConfigUpdate(
+              identityProviderId = cfg1.identityProviderId,
+              issuerUpdate = Some("issuer2"),
```
Comment (pbatko-da) [inline ledger/participant-local-store/src/test/lib/com/daml/platform/localstore/IdentityProviderConfigStoreTests.scala:248]: What if you're doing a no-op update like this:
```
IdentityProviderConfigUpdate(
              identityProviderId = cfg1.identityProviderId,
              issuerUpdate = Some("issuer1"),
            )
```
Will an error still be raised?

File: ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala [hunk 19]
```diff
@@ -415,4 +418,75 @@ object domain {
...
+      override def toDb: Option[Id] = None
+    }
+
+    final case class Id(value: Ref.LedgerString) extends IdentityProviderId {
+      override def toRequestString: String = value
+
+      override def toDb: Option[Id] = Some(this)
+    }
+
+    object Id {
```
Comment (pbatko-da) [inline ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala:457]: Do we need that object? Can we delete them and move its responsibilities to`IdentityProviderId`?

File: ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/CachedIdentityProviderConfigStore.scala [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    CaffeineCache.type,
```
Comment (pbatko-da) [inline ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/CachedIdentityProviderConfigStore.scala:45]: Is this an established pattern? Maybe instead create a custom object like `CachedIdentityProviderListSingletonKey` (or something less verbose)

File: ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala [hunk 21]
```diff
@@ -415,4 +418,75 @@ object domain {
...
+      def fromString(id: String): Either[String, IdentityProviderId.Id] = {
+        Ref.LedgerString.fromString(id).map(Id.apply)
+      }
+
+      def assertFromString(id: String) = {
```
Comment (pbatko-da) [inline ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala:462]: Missing return type

Comment (pbatko-da) [inline ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala:462]: Are these factory methods tested?

File: ledger/metrics/src/main/scala/com/daml/metrics/IdentityProviderConfigStoreMetrics.scala [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+  representative = "daml.identity_provider_config_store.<operation>",
+  groupableClass = classOf[DatabaseMetrics],
+)
+class IdentityProviderConfigStoreMetrics(
+    override val prefix: MetricName,
+    override val registry: MetricRegistry,
+) extends FactoryWithDBMetrics {
+
+  val cache = new CacheMetrics(prefix :+ "cache", registry)
+  val createIDPConfig: DatabaseMetrics = createDbMetrics("create_identity_provider_config")
```
Comment (pbatko-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/IdentityProviderConfigStoreMetrics.scala:20]:
```suggestion
  val createIdpConfig: DatabaseMetrics = createDbMetrics("create_identity_provider_config")
```
?

**Review by pbatko-da — COMMENTED** [review]
Looks good; mostly minor comments and nitpicking

File: ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala [hunk 23]
```diff
@@ -415,4 +418,75 @@ object domain {
...
+
+  sealed trait IdentityProviderId {
+    def toRequestString: String
+
+    def toDb: Option[IdentityProviderId.Id]
+  }
+
+  object IdentityProviderId {
+    final case object Default extends IdentityProviderId {
+      override def toRequestString: String = ""
```
[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala:447]: as you pointed out in the other review concern - it does not seem to be tested, so I will add a comprehensive unit test to cover this method and all others. 

the whole PR here consistent of the unused logic actually, as nothing is plugged :)

File: ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala (hunk 19 above)
Comment (skisel-da) [reply ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala:457]: There are currently 2 impl of the `IdentityProviderId` trait: `Default` and `Id` with concrete value. Not sure how to model `Default` then. `Option[IdentityProviderId]` was super confusing as could mean absence of the value (unknown) or Default case

File: ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala (hunk 21 above)
Comment (skisel-da) [reply ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala:462]: adding tests, thanks!

File: ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala [hunk 24]
```diff
@@ -415,4 +418,75 @@ object domain {
...
+        Ref.LedgerString.fromString(id).map(Id.apply)
+      }
+
+      def assertFromString(id: String) = {
+        Id(Ref.LedgerString.assertFromString(id))
+      }
+    }
+
+    def apply(identityProviderId: String): IdentityProviderId =
+      Option(identityProviderId).filter(_.nonEmpty) match {
```
[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala:468]: actually no, but "" would mean `Default`, will fix that

File: ledger/participant-integration-api/src/main/resources/db/migration/h2database/V1__Append_only_schema.sql [hunk 25]
```diff
@@ -384,6 +384,17 @@ CREATE UNIQUE INDEX participant_metering_from_to_application ON participant_mete
 
 -- NOTE: We keep participant user and party record tables independent from indexer-based tables, such that
 --       we maintain a property that they can be moved to a separate database without any extra schema changes.
+---------------------------------------------------------------------------------------------------
+-- Participant local store: identity provider configurations
+---------------------------------------------------------------------------------------------------
+CREATE TABLE participant_identity_provider_config
+(
+    identity_provider_id VARCHAR(255) PRIMARY KEY NOT NULL,
```
[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/h2database/V1__Append_only_schema.sql:392]: I was told `LedgerString` is 255:

https://github.com/digital-asset/daml/pull/15451#discussion_r1022831367

File: ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/CachedIdentityProviderConfigStore.scala (hunk 14 above)
Comment (skisel-da) [reply ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/CachedIdentityProviderConfigStore.scala:44]: This method will be heavily used for each request as we will be authenticating against all IDPs

File: ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/InMemoryIdentityProviderConfigStore.scala (hunk 16 above)
Comment (skisel-da) [reply ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/InMemoryIdentityProviderConfigStore.scala:67]: wow good catch, how come tests did not capture it

File: ledger/participant-local-store/src/test/lib/com/daml/platform/localstore/IdentityProviderConfigStoreTests.scala (hunk 18 above)
Comment (skisel-da) [reply ledger/participant-local-store/src/test/lib/com/daml/platform/localstore/IdentityProviderConfigStoreTests.scala:248]: it is currently allowed, not sure what the expectation, but let me add a test case for that

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderCheckStorageBackend.scala (hunk 4 above)
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderCheckStorageBackend.scala:9]: see PoC
UserManagementStore then later would need to check if provided identity_provider_id exists

File: ledger/participant-local-store/src/test/lib/com/daml/platform/localstore/IdentityProviderConfigStoreTests.scala (hunk 18 above)
Comment (skisel-da) [reply ledger/participant-local-store/src/test/lib/com/daml/platform/localstore/IdentityProviderConfigStoreTests.scala:248]: actually it is covered by the test and we do not expect failure in this case

File: ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala [hunk 26]
```diff
@@ -415,4 +418,75 @@ object domain {
...
+        case Some(id) => Id(Ref.LedgerString.assertFromString(id))
+        case None => Default
+      }
+
+    def fromString(identityProviderId: String): Either[String, IdentityProviderId] =
```
[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala:473]: No, removed

File: ledger/participant-integration-api/src/main/scala/platform/localstore/PersistentIdentityProviderConfigStore.scala (hunk 2 above)
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/scala/platform/localstore/PersistentIdentityProviderConfigStore.scala]: I actually considered and even had it, but in fact they're used once in this file, so there was no value to add it

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderStorageBackend.scala (hunk 9 above)
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/localstore/IdentityProviderStorageBackend.scala]: removed

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres/V126__identity_provider_config.sql [hunk 27]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE participant_identity_provider_config
+(
+    identity_provider_id VARCHAR(255) PRIMARY KEY NOT NULL,
```
[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/postgres/V126__identity_provider_config.sql:3]: how do I add a test to check ordering using collation for ledger strings? i.e. ``[A-Za-z0-9#:\-_/ ]+``.

File: ledger/participant-integration-api/src/main/scala/platform/localstore/PersistentIdentityProviderConfigStore.scala [hunk 28]
```diff
@@ -0,0 +1,172 @@
...
+  override def updateIdentityProviderConfig(update: IdentityProviderConfigUpdate)(implicit
+      loggingContext: LoggingContext
+  ): Future[Result[domain.IdentityProviderConfig]] = {
+    inTransaction(_.updateIDPConfig) { implicit connection =>
+      val id = update.identityProviderId
+      for {
+        _ <- idpConfigExists(id)
+        _ <- idpConfigIssuerDoesNotExist(update.issuerUpdate)
+        _ = {
+          update.issuerUpdate.foreach(
```
[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/scala/platform/localstore/PersistentIdentityProviderConfigStore.scala:101]: The very last call queries idp_config by id and it would give `IdentityProviderConfigNotFound` to the output, do you see a problem with this approach?

File: ledger/participant-integration-api/src/main/scala/platform/localstore/PersistentIdentityProviderConfigStore.scala [hunk 29]
```diff
@@ -0,0 +1,172 @@
...
+        s"Deleted identity provider configuration with id $id"
+      )
+    })
+  }
+
+  override def listIdentityProviderConfigs()(implicit
+      loggingContext: LoggingContext
+  ): Future[Result[Seq[domain.IdentityProviderConfig]]] = {
+    inTransaction(_.listIDPConfigs) { implicit connection =>
+      Right(backend.listIdentityProviderConfigs()(connection))
```
[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/scala/platform/localstore/PersistentIdentityProviderConfigStore.scala:88]: Implemented a limit

File: ledger/participant-integration-api/src/main/scala/platform/localstore/PersistentIdentityProviderConfigStore.scala (hunk 28 above)
[the remark this answers is not in the document]
Comment (pbatko-da) [reply ledger/participant-integration-api/src/main/scala/platform/localstore/PersistentIdentityProviderConfigStore.scala:101]: A different transaction can delete and re-create the config while this one is running.
We probably cannot be totally bullet-proof (e.g. it all can happen after the last update and before the last query which fetches the record); but it seems prudent to check if the update queries updated anything at all

File: ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala (hunk 19 above)
Comment (pbatko-da) [reply ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala:457]: I'm guessing `Default` will never be persisted to the DB and its `def toDb` method will never be used?

File: ledger/ledger-api-domain/src/test/suite/scala/com/daml/ledger/api/IdentityProviderIdSpec.scala [hunk 30]
```diff
@@ -0,0 +1,74 @@
...
+  "IdentityProviderId.Id.assertFromString" in {
+    assertThrows[IllegalArgumentException] {
+      IdentityProviderId.Id.assertFromString("")
+    }
+
+    assertThrows[IllegalArgumentException] {
+      IdentityProviderId.Id.assertFromString("a" * 256)
+    }
+
+    IdentityProviderId.Id.assertFromString("a123") shouldBe
```
Comment (mziolekda) [inline ledger/ledger-api-domain/src/test/suite/scala/com/daml/ledger/api/IdentityProviderIdSpec.scala:48]: What is this case actually testing?

File: ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/CachedIdentityProviderConfigStore.scala [hunk 31]
```diff
@@ -0,0 +1,98 @@
...
+      caffeine.Caffeine
+        .newBuilder()
+        .expireAfterWrite(Duration.ofSeconds(expiryAfterWriteInSeconds.toLong))
+        .maximumSize(1)
+        .buildAsync(
+          new FutureAsyncCacheLoader[SingletonCacheKey.type, Result[Seq[IdentityProviderConfig]]](
+            _ => delegate.listIdentityProviderConfigs()
+          )
+        ),
+      metrics.daml.identityProviderConfigStore.cache,
```
Comment (mziolekda) [inline ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/CachedIdentityProviderConfigStore.scala:59]: Sharing this metrics container is probably going to lead to confusing results.

File: ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/InMemoryIdentityProviderConfigStore.scala [hunk 32]
```diff
@@ -0,0 +1,101 @@
...
+    TrieMap[IdentityProviderId.Id, IdentityProviderConfig]()
+
+  override def createIdentityProviderConfig(identityProviderConfig: domain.IdentityProviderConfig)(
+      implicit loggingContext: LoggingContext
+  ): Future[Result[domain.IdentityProviderConfig]] = withState {
+    for {
+      _ <- checkIssuerDoNotExists(identityProviderConfig.issuer)
+      _ <- checkIdDoNotExists(identityProviderConfig.identityProviderId)
+      _ = state.put(identityProviderConfig.identityProviderId, identityProviderConfig)
+      _ <- tooManyIdentityProviderConfigs()
```
Comment (mziolekda) [inline ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/InMemoryIdentityProviderConfigStore.scala:27]: shouldn't we check for "too many configs" prior to attempting to change the state?

Comment (skisel-da) [reply ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/InMemoryIdentityProviderConfigStore.scala:27]: I was following a pattern @pbatko-da used in user management for the user rights.. I suspect transaction will be rolled back

File: ledger/participant-local-store/src/test/lib/com/daml/platform/localstore/IdentityProviderConfigStoreTests.scala [hunk 33]
```diff
@@ -0,0 +1,307 @@
...
+      issuer = UUID.randomUUID().toString,
+    )
+
+  def randomId() = {
+    val id = UUID.randomUUID().toString
+    IdentityProviderId.Id(Ref.LedgerString.assertFromString(id))
+  }
+
+  "identity provider config store" - {
+    "allows to create and load unchanged an identity provider config" in {
```
Comment (mziolekda) [inline ledger/participant-local-store/src/test/lib/com/daml/platform/localstore/IdentityProviderConfigStoreTests.scala:39]: we are not loading anything in this test, just creating

File: ledger/participant-local-store/src/test/lib/com/daml/platform/localstore/IdentityProviderConfigStoreTests.scala [hunk 34]
```diff
@@ -0,0 +1,307 @@
...
+          res1 shouldBe Right(cfg1)
+          res2 shouldBe Left(IdentityProviderConfigWithIssuerExists("issuer1"))
+        }
+      }
+    }
+
+    s"disallow to create more than 10 configs" in {
+      testIt { tested =>
+        for {
+          res1 <- Future.sequence((1 to 10).map(_ => tested.createIdentityProviderConfig(config())))
```
Comment (mziolekda) [inline ledger/participant-local-store/src/test/lib/com/daml/platform/localstore/IdentityProviderConfigStoreTests.scala:82]: Could we use the `maxIdentityProviderConfigs` here?

Comment (skisel-da) [reply ledger/participant-local-store/src/test/lib/com/daml/platform/localstore/IdentityProviderConfigStoreTests.scala:82]: sure

**Review by mziolekda — COMMENTED** [review]
Left a number of comments, but in general pretty good already!

File: ledger/ledger-api-domain/src/test/suite/scala/com/daml/ledger/api/IdentityProviderIdSpec.scala (hunk 30 above)
Comment (skisel-da) [reply ledger/ledger-api-domain/src/test/suite/scala/com/daml/ledger/api/IdentityProviderIdSpec.scala:48]: well... it's testing happy path to create Id instance... not sure how to express it differently :)

File: ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/InMemoryIdentityProviderConfigStore.scala (hunk 32 above)
Comment (pbatko-da) [reply ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/InMemoryIdentityProviderConfigStore.scala:27]: Checking before doesn't work as another transaction can insert additional configs after the check but before this transactions commits

Comment (pbatko-da) [reply ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/InMemoryIdentityProviderConfigStore.scala:27]: @skisel-da normally a transaction will be rolled back if you throw an exception and it seems it might not work in your setup when you're working with Eithers
